bindings/pyroot/pythonizations/src/TMemoryRegulator.h

```diff
@@ -39,29 +40,59 @@
 
 // ROOT
 #include "TObject.h"
-#include "TClass.h"
 
 // Stl
 #include <unordered_map>
 
 namespace PyROOT {
 
-typedef std::unordered_map<Cppyy::TCppObject_t, Cppyy::TCppType_t> ObjectMap_t;
-
-class TMemoryRegulator : public TObject {
-private:
-   static ObjectMap_t fObjectMap; // key: object address; value: object class id
+class RegulatorCleanup;
+
+/// Manages TObject-derived objects created in a PyROOT application
+///
+/// This class is responsible to keep track of the creation of the objects
+/// that need further memory management within ROOT. The `ClearProxiedObjects`
+/// function is only called at PyROOT shutdown time. The `CallCppyyRecursiveRemove`
+/// is called as part of the global list of cleanups object destruction.
+///
+/// This class is intentionally not derived from TObject. See the
+/// `PyROOT::RegulatorCleanup` class for more info.
+///
+/// \note This class is made thread-safe by having a static *thread-local*
+///       map data member to keep track of the registered objects.
+class TMemoryRegulator final {
+   using ObjectMap_t = std::unordered_map<Cppyy::TCppObject_t, Cppyy::TCppType_t>;
+
+   // The object is thread-local because it can happen that we call into
+   // C++ code (from the PyROOT CPython extension, from CPyCppyy or from cling)
+   // from different Python threads. A notable example is within a distributed
+   // RDataFrame application running on Dask.
+   static thread_local ObjectMap_t fgObjectMap; // key: object address; value: object class id
```

Comment:
If the `TMemoryRegulator` is only used as a member of the `RegulatorCleanup`, which in turn is only instantiated as `thread_local PyROOT::RegulatorCleanup m` in PyROOTWrapper.cxx, isn't it redundant to have the members of the `TMemoryRegulator` thread_local?

Reply:
I also thought about that. Then I saw that `fgObjectMap` is used in the static functions `TMemoryRegulator::RegisterHook` and `TMemoryRegulator::UnregisterHook`. Thus the map needs to be at least static so those functions can use it without taking it as an argument (the signature is fixed by cppyy's MemoryRegulator). If I left only `static fgObjectMap` without the `thread_local` part it would probably defeat the purpose because there would be one memory regulator per thread but only one static map.

Reply:
Ah, ok thanks! Well then the next questions would be why isn't the `CPyCppyy::MemoryRegulator` thread local too, but apparently it wasn't necessary to make things work on Ubuntu 24 :+1:

Reply:
Actually, come think of it, there's a very easy way to implement your suggestion, I'm uploading a new commit.